### PR TITLE
Rewrite haplotype construction methods in PDHapComputationEngine

### PR DESCRIPTION
**Pull request by davidbenjamin**
@jamesemery Here's another fun one, again no change in output but significant refactoring of `constructHaplotypeFromVariants` and `createNewPDHaplotypeFromEvents`.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 1]
```diff
@@ -550,150 +550,100 @@ private static byte[] basesAfterFirst(final Allele altAllele) {
...
-                basesToAdd = refAllele.length() >= altAllele.length() ? refAllele.getBases() : altAllele.getBases();
-            }
+            final int altRefLengthDiff = altAllele.length() - refAllele.length();
 
-            // Remove anchor base if necessary
-            if (event.isIndel() && !includeRefBaseForIndel) {
-                basesToAdd = Arrays.copyOfRange(basesToAdd, 1, basesToAdd.length);
-            }
+            boolean isInsertion = altRefLengthDiff > 0; // If its an insertion we flip to "ADD" the bases to the ref.
+            final boolean isEvent = event.getStart() == eventWithVariant.getStart();
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:604]: using evetn.getStart() seems dangerous (the old code is using it too i thought i wasn't) seems to preclude compound events where a snp and indel start on the same position but only one is determined? I would need to dig over the code again to see if its possible for those to happen but this equality assertion gives me some pause... I know compound UNDETERMINED events are allowed and perhaps thats why this isn't causing problems

**Review by jamesemery — COMMENTED** [review]
I need to keep staring at this.... you should take a look at the unit tests for this class and consider if its worth adding some more test cases

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 2]
```diff
@@ -550,150 +550,100 @@ private static byte[] basesAfterFirst(final Allele altAllele) {
...
      */
     @VisibleForTesting
-    //TODO When we implement JointDetection we will need to allow multiple eventWithVariants to be present...
-    static PartiallyDeterminedHaplotype createNewPDHaplotypeFromEvents(final Haplotype base, final Event eventWithVariant, final boolean useRef, final List<Event> constituentEvents) {
-        //ASSERT that the base is ref and cool
-        if (!base.isReference() || base.getCigar().numCigarElements() > 1) {
-            throw new RuntimeException("This is not a valid base haplotype for construction");
-        }
+    //TODO When we implement JointDetection we will need to allow multiple eventWithVariants to be prsent...
+    static PartiallyDeterminedHaplotype createNewPDHaplotypeFromEvents(final Haplotype refHap, final Event eventWithVariant, final boolean useRef, final List<Event> constituentEvents) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:564]: I'm going to leave that for next week or so in order to cleanly separate pure refactoring PRs from new functionality PRs.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 3]
```diff
@@ -550,150 +550,100 @@ private static byte[] basesAfterFirst(final Allele altAllele) {
...
-            }
-
-            // Add the cigar for bases we skip over
-            if (intermediateRefEndPosition - intermediateRefStartPosition > 0) {
-                runningCigar.add(new CigarElement(intermediateRefEndPosition - intermediateRefStartPosition, CigarOperator.M));
-            }
-
-            // Include the ref base for indel if the base immediately proceeding this event is not already tracked
-            boolean includeRefBaseForIndel = event.isIndel() && (intermediateRefStartPosition <= intermediateRefEndPosition);
+            Utils.validate(basesBeforeNextEvent >= 0, () -> "PD event list: " + constituentEvents + " is out of order.");
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:597]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 4]
```diff
@@ -550,150 +550,100 @@ private static byte[] basesAfterFirst(final Allele altAllele) {
...
+            newHapBases.put(ArrayUtils.subarray(refBasesToAddTo, lastPositionAdded - refStart, actualStart - refStart)); // bases before the variant
+            pdBytes.put(new byte[actualStart - lastPositionAdded]); // bases before the variant -- all zeroes
+
+            // Now add event bases.  If this is the blessed variant, add the ref or alt as appropriate
+            // Otherwise make sure we are adding the longest allele (for indels) or the ref allele for snps.
+            final Allele alleleToUse = (isEvent && useRef) || (!isEvent && altRefLengthDiff <= 0) ? refAllele : altAllele;
+            final byte[] basesToAdd = event.isIndel() ? basesAfterFirst(alleleToUse) : alleleToUse.getBases();
+            newHapBases.put(basesToAdd); // refbases added
+            pdBytes.put(isEvent ? new byte[basesToAdd.length] :
+                    PartiallyDeterminedHaplotype.getPDBytesForHaplotypes(isInsertion ? altAllele : refAllele, isInsertion ? refAllele : altAllele)); // refbases added
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:632]: I extracted two methods: `Event::longerAllele()` and `Event::shorterAllele()`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:604]: Now using `Event::equals`

**Comment by davidbenjamin** [comment]
Back to you, @jamesemery

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 5]
```diff
@@ -562,150 +550,101 @@ public static Haplotype constructHaplotypeFromVariants(final Haplotype refHap, f
...
-        newHaplotypeBasees = ArrayUtils.addAll(newHaplotypeBasees, ArrayUtils.subarray(refBasesToAddTo, refStartIndex, refBasesToAddTo.length));
-        pdBytes = ArrayUtils.addAll(pdBytes, new byte[refBasesToAddTo.length - refStartIndex]);
+        int refStartIndex = lastPositionAdded - refStart;
+        newHapBases.put(ArrayUtils.subarray(refBasesToAddTo, refStartIndex, refBasesToAddTo.length));
+        pdBytes.put(new byte[refBasesToAddTo.length - refStartIndex]);
         runningCigar.add(new CigarElement(refBasesToAddTo.length - refStartIndex, CigarOperator.M));
 
         return new PartiallyDeterminedHaplotype(
-                new Haplotype(newHaplotypeBasees, false, base.getGenomeLocation(), runningCigar.make()),
+                new Haplotype(ArrayUtils.subarray(newHapBases.array(), 0, newHapBases.position()), false, refHap.getGenomeLocation(), runningCigar.make()),
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:639]: can you put a check into this constructor that the pdhapbases and the haplotype bases match in length?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 6]
```diff
@@ -479,81 +477,71 @@ private static boolean constructArtificialHaplotypeAndTestEquivalentEvents(Haplo
...
         runningCigar.add(new CigarElement(refbases.length - refStartIndex, CigarOperator.M));
 
-        final Haplotype outHaplotype = new Haplotype(newRefBases, false, refHap.getGenomeLocation(), runningCigar.make());
+        final Haplotype result = new Haplotype(newHapBases.array(), false, refHap, runningCigar.make());
         if (setEventMap) {
-            EventMap.buildEventMapsForHaplotypes(Collections.singletonList(outHaplotype), refHap.getBases(), refHap.getGenomeLocation(), false,0);
+            EventMap.buildEventMapsForHaplotypes(List.of(result), refHap.getBases(), refHap, false,0);
             // NOTE: we set this AFTER generating the event maps because hte event map code above is being generated from the ref hap so this offset will cause out of bounds errors
-            outHaplotype.setAlignmentStartHapwrtRef(refHap.getAlignmentStartHapwrtRef()); //TODO better logic here
+            result.setAlignmentStartHapwrtRef(refHap.getAlignmentStartHapwrtRef()); //TODO better logic here
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:533]: can you get the typos and possibly remove this todos?

**Review by jamesemery — APPROVED** [review]
looks good, very minor comments but otherwise i see no reason to keep blocking this

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 6 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:533]: done x 2

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 5 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:639]: done

**Pull request merged**